### ISSUE TITLE
runtime error: member access within address 0x604000001110 with insufficient space for an object of type 'JsonVariant' 

### ISSUE THREAD

**Issue by mrc0mmand** [systemd/systemd#13020]
**systemd version the issue has been seen with**
> latest master

**Used distribution**
 > Arch Linux

In https://github.com/systemd/systemd-centos-ci/pull/144 (not only) TEST-01-BASIC started failing due to an UBSan error:

```
../src/shared/json.c:1671:20: runtime error: member access within address 0x604000001110 with insufficient space for an object of type 'JsonVariant' (aka 'struct JsonVariant')
0x604000001110: note: pointer points here
 02 00 80 5b  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00
              ^ 
    #0 0x7fe3a6201791 in json_variant_copy /build/build/../src/shared/json.c:1671:20
    #1 0x7fe3a6200a06 in json_variant_set_source /build/build/../src/shared/json.c:1744:13
    #2 0x7fe3a61f54f5 in json_parse_internal /build/build/../src/shared/json.c:2593:32
    #3 0x7fe3a61f3b7e in json_parse /build/build/../src/shared/json.c:2622:16
    #4 0x7fe3a624f50a in varlink_parse_message /build/build/../src/shared/varlink.c:581:13
    #5 0x7fe3a624c8e3 in varlink_process /build/build/../src/shared/varlink.c:926:13
    #6 0x7fe3a625a535 in defer_callback /build/build/../src/shared/varlink.c:1764:16
    #7 0x7fe3a6500f50 in source_dispatch /build/build/../src/libsystemd/sd-event/sd-event.c
    #8 0x7fe3a64ff880 in sd_event_dispatch /build/build/../src/libsystemd/sd-event/sd-event.c:3241:21
    #9 0x7fe3a650215c in sd_event_run /build/build/../src/libsystemd/sd-event/sd-event.c:3299:21
    #10 0x55c46d7fe85f in main /build/build/../src/journal/journald.c:92:21
    #11 0x7fe3a5b53ce2 in __libc_start_main (/usr/lib/libc.so.6+0x23ce2)
    #12 0x55c46d7fe38d in _start (/usr/lib/systemd/systemd-journald+0x4738d)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../src/shared/json.c:1671:20 in 
==49==WARNING: Symbolizer was blocked from starting itself!

=================================================================
==49==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 67584 byte(s) in 1 object(s) allocated from:
    #0 0x7fb2f36c3919  (/usr/lib/clang/8.0.0/lib/linux/libclang_rt.asan-x86_64.so+0x123919)
    #1 0x7fb2efabbaf3  (/usr/bin/../lib/libLLVM-8.so+0x9f3af3)
    #2 0x3f403ef1f  (<unknown module>)
    #3 0x7ffd6e48edcd  ([stack]+0x20dcd)

SUMMARY: AddressSanitizer: 67584 byte(s) leaked in 1 allocation(s).
```

I'm not sure about the LSan error, will look into it.

**Comment by poettering** [comment]
Hmm i wonder if we are really doing something undefined here, or if ubsan is stupid...

So the structure ends in a union, and that union can either be relatively short or long. When allocating the whole structure we only allocate the short versin since we know we won't need the long version. This means we knowingly allocate a structure shorter than its definition. Which should normally be fine, as long as we never access the end. But maybe ubsan is right and the C compiler has the liberty to access to the "unused" areas of a union if it likes...

**Comment by poettering** [comment]
I prepped a fix for this in #13021. At least I hope that fixes it. PTAL.

**Comment by mrc0mmand** [comment]
In my attempts to reproduce this locally I accidentally managed to trigger this one:

```
# build/test-alloc-util
../src/test/test-alloc-util.c:69:9: runtime error: load of address 0x60200000003c with insufficient space for an object of type 'int'
0x60200000003c: note: pointer points here
  03 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00
              ^ 
    #0 0x40307c in test_memdup_multiply_and_greedy_realloc ../src/test/test-alloc-util.c:69
    #1 0x403a7c in main ../src/test/test-alloc-util.c:155
    #2 0x7f9b48a38412 in __libc_start_main (/lib64/libc.so.6+0x24412)
    #3 0x4012ad in _start (/root/systemd/build/test-alloc-util+0x4012ad)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../src/test/test-alloc-util.c:69:9 in
```
